### ISSUE TITLE
How to average sub-words embeddings to obtain word embeddings?

### ISSUE THREAD

**Issue by speedcell4** [huggingface/transformers#2079]
Hi~

How to average sub-words embeddings to obtain word embeddings?
I only want word-level embedding instead of sub-word-level, how can I get them?

Is there any tokenizer that provides a method that can output the index/mask of sub-words or something?

**Comment by karajan1001** [comment]
You may use the word as the input and make the sentence embedding as the word embedding.
for example, input is 
"puppeteer"
tokens as
'[CLS]', 'puppet', '##eer', '[SEP]'
and then get embedding of this tokens list output.

**Comment by zhoudoufu** [comment]
I have similar usage as well, I did a simple experiment, and observe that the subword embedding when input a whole sentence, the cosine similarity tends to above 90%. So that sum and average subwords' embedding doesn't change much.

**Comment by glicerico** [comment]
Take a look at how bert-sense does it :)
https://github.com/uhh-lt/bert-sense/blob/bfecb3c0e677d36ccfab4e2131ef9183995efaef/BERT_Model.py#L342

**Comment by mathshangw** [comment]
Excuse me can someone solve it ?

**Comment by mathshangw** [comment]
Excuse me did some one solve it ?

**Comment by speedcell4** [comment]
@mathshangw I turn to use the `TransformerWordEmbeddings` of `flair` library to handle this. Here is an [example](https://github.com/flairNLP/flair/blob/master/examples/ner/run_ner.py#L119).

**Comment by speedcell4** [comment]
I found a way to obtain the subtoken mask.

There is an argument called `return_offsets_mapping`, when you pass the tokenized sequence to the tokenizer, the returned offsets mapping records the start position of each token instead of the entire sentence, for example.

```python
tokens: list[str] = 'this is a niceing work'.split()

tokenizer.tokenize(tokens, add_special_tokens=True)
# ['▁this', '▁is', '▁a', '▁nice', 'ing', '▁work', '</s>', 'en_XX']

tokens = tokenizer(tokens, add_special_tokens=True, is_split_into_words=True,
            return_offsets_mapping=True, return_tensors='pt')
# {
#    'input_ids': tensor([[903, 83, 10, 26267, 214, 4488, 2, 250004]]),
#    'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1]]),
#    'offset_mapping': tensor([[[0, 4],
#                               [0, 2],
#                               [0, 1],
#                               [0, 4],
#                               [4, 7],
#                               [0, 4],
#                               [0, 0],
#                               [0, 0]]])}

subtoken_mask = tokens['offset_mapping'][..., 0] != 0
# tensor([[False, False, False, False,  True, False, False, False]])
```

Forget about the weird word *nicing*, I just want to get some sub-tokens. Now, simply checking if this token starts from the beginning of the given word, we can tell if it is a sub-token.

Hope this is helpful to you guys.
